### PR TITLE
fix: virtualize payments, purses, ledger

### PR DESCRIPTION
**Pull request by erights**
Virtualize payment, paymentLedger, and purse,depositFacet together as a multi-faceted unit.

This does not virtualize everything that will need to be durable to support upgrade.

This does not even virtualize everything of high arity, needed for scaling, as it does not virtualize the per-purse notifier.

### CODE REVIEW

**Comment by erights** [comment]
Assuming it passes CI enough, this is now ready for review. Thanks!

File: packages/ERTP/src/payment.js
Comment (turadg) [inline packages/ERTP/src/payment.js:3]: this deep import is unfortunate. I'll prioritize https://github.com/Agoric/agoric-sdk/issues/4255

File: packages/ERTP/src/purse.js
Comment (turadg) [inline packages/ERTP/src/purse.js:13]: Shouldn't the allegedName include a description of the kind?

What is the kind being defined here? From `makePurseKit` I read this as a "purse kit" kind, but I'm surprised to see a kit be a durable thing with state. I thought we used kit as a collection of related functionality.

It looks like this was made to have a place outside `purse` to the deposit facet but I don't see the motivation for that. Especially since `purse` still has `getDepositFacet`.

**Comment by erights** [comment]
Solving a CI failure led me down a rabbit hold, resulting in https://github.com/Agoric/agoric-sdk/pull/4618/commits/981bc1e95bf6a4914398def26ad3b450b38690cb which we should definitely discuss before merging. I would like to avoid merging this as part of this PR.

However, without it, the generated file ui-components/compiled/test/components/test-NatAmountInput.js fails, apparently because its imports include ERTP, which now depends on storeModule, which depends on the global variable `VatData`. However, apparently this generated file is run in an environment in which no one installs `VatData`.

See #4255

File: packages/ERTP/src/payment.js
Comment (erights) [reply packages/ERTP/src/payment.js:3]: Thanks! See also my comment https://github.com/Agoric/agoric-sdk/pull/4618#issuecomment-1074462354 below.

File: packages/ERTP/src/purse.js
Comment (erights) [reply packages/ERTP/src/purse.js:13]: > Shouldn't the allegedName include a description of the kind?

The "kind" is the pair of the `purse` and `depositFacet`, of which only the `purse` is immediately exposed.

Since we're calling (the maker returned by) `defineKind` with (a function that returns) a record with both `purse` and `depositFacets`, and only these objects are declared `Far`, `defineKind` will call `Far` with farNames of, respectively 
```js
`${allegedName} purse`
```
and
```js
`${allegedName} depositFacet`
```
which, somewhat fortuitously, actually mimics its previous behavior.

Comment (erights) [reply packages/ERTP/src/purse.js:13]: > It looks like this was made to have a place outside purse to the deposit facet but I don't see the motivation for that. Especially since purse still has getDepositFacet.

The `purse` and `depositFacet` need each other, but both are exposed to the outside world as independent `Far` objects. However, between the two of them, only one state record is needed. This solution doesn't pay any storage in the `state` record for the `depositFacet` because it only needs to exist at runtime, reactivated at the same time the purse is. But it still preserves the identity of the per-purse depositFacet as a Far object.

Do you have another solution in mind?

**Comment by erights** [comment]
Yay! CI passes! But only with the https://github.com/Agoric/agoric-sdk/commit/981bc1e95bf6a4914398def26ad3b450b38690cb kludge. Is there a better way?

File: packages/ERTP/src/purse.js
Comment (turadg) [reply packages/ERTP/src/purse.js:13]: > Do you have another solution in mind?

Not yet; trying to understand this:

> both are exposed to the outside world as independent Far objects

In `master`, `depositFacet` is exposed only via `purse.getDepositFacet()`. Why the change to expose it as an independent Far object?  https://github.com/Agoric/agoric-sdk/blob/45723b48823ed59c647d4b1f4b4d7ae16c421209/packages/ERTP/src/purse.js#L35-L43

Comment (erights) [reply packages/ERTP/src/purse.js:13]: The `depositFacet` is exposed *only* by `purse.getDepositFacet()`. But since it is exposed, it must be a far object. Since it has the same cardinality as purses, it must also be virtualized. But since it has no state independent of its purse, it need not be separately virtualized. It can just be virtualized as one of the cohort pair activated for reviving that state record.

Comment (warner) [reply packages/ERTP/src/purse.js:13]: IIRC, we figured that `getDepositFacet()` *could* create a new facet on demand, which would just wrap the Purse and expose a deposit method, but 1: it made a good candidate for practicing the multi-facet objects, and 2: there may have been some maintain-a-constant-identity behavior that was easier/only-possible to achieve with the multi-facet approach.

File: packages/governance/test/unitTests/test-buildParamManager.js
Comment (warner) [inline packages/governance/test/unitTests/test-buildParamManager.js:217]: Huh, why did this need to change?

File: packages/SwingSet/src/storeModule.js
Comment (warner) [inline packages/SwingSet/src/storeModule.js:25]: this makes me nervous, but it sounds like other people have a plan to rationalize access to `VatData`, so I'm not going to worry about it

**Review by warner — APPROVED** [review]
This seems like a good start. I'm not excited about the VatData thing, but I figure we'll find a solution for that sooner or later.

While reviewing this, I realized we have a GC/determinism problem, which I wrote up in #4892. It's not specific to this conversion though, so it shouldn't hold up this PR.

File: packages/ERTP/src/purse.js
Comment (turadg) [reply packages/ERTP/src/purse.js:13]: Ah. Since this code is meant to serve as an example I'd ask that some of the above be documented in it. If I've got this right,
```
export const makePurseMaker = (allegedName, assetKind, brand, purseMethods) => {
  // - This kind is a pair of purse and depositFacet that have a 1:1 correspondence.
  // - They are virtualized together to share a single state record.
  // - An alternative design considered was to have this return a Purse alone
  //   that created depositFacet as needed but this approach ensures a constant
  //   identity for the facet and exercises the multi-faceted object style.
  const makePurseKit = defineKind(
```

File: packages/ERTP/test/unitTests/test-issuerObj.js
Comment (turadg) [inline packages/ERTP/test/unitTests/test-issuerObj.js]: there's no `^` so it the leading wildcard necessary?

File: packages/governance/test/unitTests/test-buildParamManager.js
Comment (turadg) [reply packages/governance/test/unitTests/test-buildParamManager.js:217]: Surprising to me too. Running the test without this change,
```
  Values are deeply equal to each other, but they are not the same:

  [
    {
      description: 'simple',
      handle: Object @Alleged: InvitationHandle {},
      installation: Object @Alleged: Installation {
        getBundle: Function getBundle {},
      },
      instance: Object @Alleged: InstanceHandle {},
    },
  ]

  › packages/governance/test/unitTests/test-buildParamManager.js:216:5
```
Comparing each named key, the are the same value,
```
  for (const k of Object.keys(invitationActualAmount[0])) {
    t.is(invitationActualAmount[0][k], invitationAmount.value[0][k]);
  }
```
But the object containing them is different. Do you know why that is @erights ?

File: packages/ERTP/src/purse.js
Comment (erights) [reply packages/ERTP/src/purse.js:13]: Done, thanks.

File: packages/governance/test/unitTests/test-buildParamManager.js
Comment (erights) [reply packages/governance/test/unitTests/test-buildParamManager.js:217]: > But the object containing them is different. Do you know why that is @erights ?

Yes. The amount is a copyRecord, which is pass-by-copy. It is serialized into persistent storage, and restored from persistent storage by unserializing. This preserves `keyEQ` equivalence which happens to be similar enough to AVA's `deepEqual` equivalence. We do not preserve JavaScript object identity for pass-by-copy objects.

Comment (erights) [reply packages/governance/test/unitTests/test-buildParamManager.js:217]: I added a `keyEQ` test to clarify.

File: packages/SwingSet/src/storeModule.js
Comment (erights) [reply packages/SwingSet/src/storeModule.js:25]: Thanks. Resolving.

File: packages/governance/test/unitTests/test-buildParamManager.js
Comment (turadg) [reply packages/governance/test/unitTests/test-buildParamManager.js:217]: Got it. Tests also serve as documentation of how things are expected to work, so would it be more accurate to use `keyEQ` here?

```suggestion
  t.true(keyEQ(invitationActualAmount, invitationAmount.value));
```

File: packages/ERTP/test/unitTests/test-issuerObj.js
Comment (erights) [reply packages/ERTP/test/unitTests/test-issuerObj.js]: Done.

**Review by turadg — APPROVED** [review]
Golden!

**Pull request merged**
